### PR TITLE
Replace std.load/std.store by affine.load/affine.store

### PR DESCRIPTION
**Pull request by tungld**
This patch is to replace all uses of std.load/std.store by affine.load/affine.store, except the following circumstances:
- Unknown dimensions in case of broadcasting operations where there is currently a lack of computing induction variables in terms of affine-map.

### CODE REVIEW

**Comment by doru1004** [comment]
@tungld could you give us a few more details that motivates this change? It looks like a very involved patch and I would like to better understand the benefits.

Since this patch changes a lot of places in the code, could this be done in a more modular manner in which, instead of changing the output code of every operation you could introduce a pass which transforms the Standard load/stores into Affine load/store wherever it's safe to do so?

The even bigger picture question is: does MLIR already have such a pass which we could invoke?

**Comment by tungld** [comment]
@doru1004 MLIR has a pass named `affine-loop-fusion` to fuse `affine.for` loops. Since, we finally emit `affine.for` from `krnl.iterate`, we would like to enable the pass for optimizing `affine.for` loops. To make the pass work, we need to use affine.load and affine.store instead of std.load and std.store, which motivates this patch. Also, I am not clear about the benefit of using std.load/std.store inside a `affine.for`, except where we cannot use affine.load/affine.store instead.

Actually, we did the same changes for Conv and Unary elementwise op in a previous patch, but we temporarily disabled `affine-loop-fusion` because it had a bug (I created a patch fix [the bug ](https://reviews.llvm.org/D82158) in MLIR and it will be merged soon). 

> Since this patch changes a lot of places in the code, could this be done in a more modular manner in which, instead of changing the output code of every operation you could introduce a pass which transforms the Standard load/stores into Affine load/store wherever it's safe to do so? This will allow us to turn this transformation on and off if necessary.

It would be great if we can do so though I feel it likes a reverse direction. The most non-trivial part is to derive an affine-map to be used in affine.load/store. Being said that, it is quite complicated for some ops like: pooling, broadcasting ops where access indices are computed using `addi`, `muli`, `subi`, `select`, etc. I think it is not so difficult for users to use affine.load/store when lowering an ONNX op to Krnl dialect.
It seems the effectiveness of affine.load and affine.store will be the same as std.load and std.store if we don't enable optimizations in Affine dialect. So it is basically turned off if we don't enable, say, `affine-loop-fusion`.

> The even bigger picture question is: does MLIR already have such a pass which we could invoke?

It doesn't have as far as I know.

**Comment by tjingrant** [comment]
I think this patch is a step in the right direction - replacing std.load/store with affine.load/store is to instill extra information/optimization hint into the IR, since it informs optimizers that access patterns are affine expressions of loop indices. The combination of affine.for and affine.load/store will enable more built-in optimization mechanisms down the road, which will catch optimization opportunities one may miss or don't want to be bothered with at higher levels of abstraction.

I'm fine with this patch as long as @doru1004 's concerns are addressed. I can also help answer his question -
> introduce a pass which transforms the Standard load/stores into Affine load/store wherever it's safe to do so? This will allow us to turn this transformation on and off if necessary

This is very difficult to do. What you are suggesting is a tool to automatically extract affine SCoP (static control parts) from the universe of possible IRs/program source code. The difficulty underpinning this task is exactly the reason why we have Affine dialect to begin with - we want to mark regions/components of IR/code as affine from the very beginning when they are generated, and keep this information for as long as reasonable so that code optimizers/transformers can make use of this information.

**Comment by doru1004** [comment]
@tjingrant @tungld thanks a lot for the detailed explanations, I just wanted to make sure we are not overlooking other methods of doing this which would end up being simpler. When we emit code for a particular operation, how do we know that we have used affine load/store correctly? Can we record some developer guidelines for that? Otherwise this patch LG.

**Comment by tjingrant** [comment]
>  When we emit code for a particular operation, how do we know that we have used affine load/store correctly? 

I believe there's a correct-by-construction mechanism in place. I.e., you won't be able to create an affine load/store unless its valid. Since it only admits operands that are affine functions of loop induction variables (so they're either induction variables or the output of affine.apply). I'm not sure how air-tight this mechanism is, but seems to go a long way to ensure correctness.

> Can we record some developer guidelines for that? 

Let me follow up with another patch to add just that.

File: test/mlir/onnx/onnx_lowering.mlir [hunk 1]
```diff
@@ -38,10 +38,10 @@ func @test_elementwise_op_with_scalar_values_2(%arg0 : tensor<f32>, %arg1 : tens
 
   // CHECK-LABEL: test_elementwise_op_with_scalar_values_2
   // CHECK: [[RES:%.+]] = alloc() : memref<f32>
-  // CHECK: [[LOAD1:%.+]] = load %arg0[] : memref<f32>
-  // CHECK: [[LOAD2:%.+]] = load %arg1[] : memref<f32>
+  // CHECK: [[LOAD1:%.+]] = affine.load %arg0[] : memref<f32>
```
Comment (max-ku) [inline test/mlir/onnx/onnx_lowering.mlir:41]: Is there a benefit in having affine loads/stores for zero dim memrefs? I do not see one, maybe I miss something?

**Pull request merged**